### ISSUE TITLE
Implement FSDP, tensor-parallel, and pipeline-parallel execution

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#99]
> Filed against the `feature_request.yml` template. Category: **Optimizer / Loss Function / Autograd** (training runtime).

## Problem

Data parallelism replicates the model on every device. It does nothing for a model that does not fit on one. The planning layer already covers the three strategies that do, and none of them executes.

`crates/incin-core/src/dist/` contains `fsdp.rs`, `tensor_parallel.rs`, and `pipeline.rs`. Their public exports are not stubs:

```text
fsdp      FsdpPlan, FsdpPlanBuilder, FsdpParameterDescriptor, FsdpMemoryReport, ZeROStage
pipeline  PipelinePlan, PipelineSchedule, GPipe, OneForwardOneBackward, PipelineClock,
          ActivationCheckpoint, PipelineTransfer, validate_microbatches
placement Sharded, Replicated, Partial, Sum/Mean/Max/Min/Prod, PlacementAxis, PipelineStage
plan      HybridPlanner, StrategyCandidate, StrategyRejection, PlanObjective, MemoryLimit,
          ShardEvidence, CommunicationEvidence, preflight
```

`docs/plan/roadmap.md` records the whole `dist` theme as "Planning and placement layer only; there is no distributed execution path." A user can express and validate a tensor-parallel plan for a model too large for one device, receive a `ShardEvidence` explaining why it is valid, and then not run it.

## Proposed API sketch

```rust
use incin::prelude::*;
use incin_core::dist::{MeshSpec, Data, TensorParallel, Pipeline, ZeROStage};
use incin_core::typenum::{U1, U2, U4};

// ── FSDP: shard parameters across the data-parallel axis ──
type Mesh = MeshSpec<Data<U4>>;

let plan = Trainer::plan()
    .devices(DeviceSet::cuda([0, 1, 2, 3])?)
    .mesh::<Mesh>()
    .sharding(ShardingSpec::Fsdp {                       // proposed
        stage: ZeROStage::Three,                         // params + grads + optimizer state
        checkpoint: ActivationCheckpoint::EveryLayer,
    })
    .build()?;

// ── Tensor parallel: partition the model across a TP axis ──
type TpMesh = MeshSpec<Data<U1>, TensorParallel<U2>>;

// Placement is already type-level today; the missing half is lowering it.
let w: Tensor<s![4096, 4096], Backend, f32, Grad> = /* ... */;
let sharded: Placed<Sharded<Axis<1>>, _> = w.place::<TpMesh, Sharded<Axis<1>>>()?;   // proposed
let y = sharded.matmul(&x)?;   // proposed: inserts the boundary all-reduce from the plan

// ── Pipeline: stages plus microbatching ──
type PpMesh = MeshSpec<Data<U1>, TensorParallel<U1>, Pipeline<U4>>;

let plan = Trainer::plan()
    .mesh::<PpMesh>()
    .schedule(PipelineSchedule::OneForwardOneBackward { microbatches: 8 })  // proposed
    .build()?;
```

The point of every sketch above: **the plan types already exist**. What is proposed is the `Trainer` accepting them and the lowering that makes them run.

## Shape and type-system invariants

This is where the existing work pays off, and the execution path must consume these proofs rather than re-derive them.

| Invariant | Already proven by | Runtime obligation |
|---|---|---|
| Mesh topology valid | `ValidMesh`, compile time | none |
| Shard divisibility (`4096 % TP == 0`) | Type-level placement rules where the extent is static | Checked for `Dyn` |
| Collective compatibility at partition boundaries | `CollectiveDescriptor`, `preflight` | Transport actually supports it |
| Every rank agrees on the plan | `AgreedPlan`, `TopologyFingerprint` | Digest comparison at rendezvous |
| Microbatch count valid | `validate_microbatches` | none |
| Memory fits | `FsdpMemoryReport`, `MemoryLimit` | Actual device memory queried |

The fail-stop contract carries over unchanged: **a plan that cannot be executed correctly must refuse, not approximate.**

## Background

**Tiering, because these are three different execution models that share only their planning.**

- **FSDP first.** It is data parallelism plus parameter sharding: all-gather a layer's parameters before use, free them after, reduce-scatter the gradients. It reuses #97's collective path almost entirely, and gives the largest memory reduction for the least new machinery. `ZeROStage` already distinguishes stages 1 through 3, so the plan can express which of optimizer state, gradients, and parameters are sharded.
- **Tensor parallel second.** Requires a partitioned matmul with an all-reduce or all-gather at the boundary. The divisibility proofs already performed are what make lowering safe; the execution needs the partitioned operation on the backend, which is why it depends on the CUDA matmul work.
- **Pipeline parallel third.** A scheduling problem rather than a collective one: microbatch sequencing, activation handoff between stages, and the bubble. `GPipe` and `OneForwardOneBackward` are both already modelled, and they share the least with the other two.

**Activation checkpointing and sharding interact, and getting it wrong looks like a memory bug.** Under FSDP a layer's parameters are freed after the forward pass. Recomputation during backward must **re-gather** them. An implementation that recomputes against freed or stale parameters produces wrong gradients silently, not a crash. `ActivationCheckpoint` already exists in the pipeline module and this interaction needs an explicit test rather than a documentation note.

**Reduce-scatter is not all-reduce.** Under FSDP each rank keeps only its shard of the gradient, so the collective is a reduce-scatter, not the all-reduce #97 uses. Reusing the wrong one produces correct-looking results and N times the memory.

## How this interoperates

- **#97** establishes the collective plumbing, bucketing, and fail-stop machinery. This is the layer above it, and should not land first.
- **#98** is reused unchanged for the data-parallel axis. Once a mesh has a TP or PP degree, the sampler must shard over the **data-parallel degree**, not the world size, or ranks within a TP group receive different data and the partitioned matmul is meaningless. That is a small change to #98's input and worth stating here so it is not discovered late.
- **#85 and #90** are prerequisites for tensor parallelism: a partitioned matmul needs a batched, dtype-parametric matmul to partition.
- **#82** fixes the `dist2-network` and `multinode` job resolution.
- **`distributed-reference`** makes all three strategies testable in-process, which is what keeps this out of the hardware-only lane.
- **Checkpointing:** single-node FSDP can save by gathering full parameters to rank 0, so it is usable with **#12** (sharded Safetensors index read) alone. **#13** (streamed checkpoint placement with bounded host memory) is what a sharded checkpoint *write* across hosts needs, and is deliberately outside this milestone. A model too large to gather onto one rank is therefore out of 0.2.0 scope, which is the scope line recorded in tracker #105.
- **`incin-telemetry`** should expose the pipeline bubble and per-stage occupancy; a pipeline schedule with no visibility into its own idle time cannot be tuned.

## Acceptance criteria

- FSDP trains a model whose parameters do not fit on a single device, and matches the unsharded loss trajectory within documented tolerance.
- FSDP uses reduce-scatter for gradients, verified by a memory measurement rather than by inspection.
- Activation checkpointing composes with FSDP, covered by a test that would fail if recomputation used stale parameters.
- A tensor-parallel plan executes across two devices and matches the single-device reference.
- A pipeline-parallel plan executes with microbatching and matches the single-device reference, with the bubble reported.
- Each strategy runs under the reference transport in-process in `ci.yml`.
- `docs/book/src/distributed.md` no longer describes the feature as planning-only for the strategies that have landed.

## Alternatives

- **Ship FSDP alone**, deferring tensor and pipeline parallelism. A reasonable smaller scope, and the tiering above is written so this can be decided later without rework. It should be a recorded decision in tracker #105 rather than the result of running out of time.
- **ZeRO stage 2 only** (shard gradients and optimizer state, replicate parameters). Much simpler, no parameter all-gather, and gives most of the memory saving for optimizer-state-dominated models. Worth landing first inside the FSDP tier.
- **Skip pipeline parallelism entirely.** It has the worst compute efficiency of the three because of the bubble, and it is the only one that helps when a single *layer* fits but the depth does not. Defensible to cut; not defensible to leave silently unimplemented behind a complete planning surface.

## Additional context

Relevant code: `crates/incin-core/src/dist/fsdp.rs` (`FsdpPlan`, `ZeROStage`, `FsdpMemoryReport`), `crates/incin-core/src/dist/tensor_parallel.rs`, `crates/incin-core/src/dist/pipeline.rs` (`GPipe`, `OneForwardOneBackward`, `ActivationCheckpoint`, `PipelineClock`), `crates/incin-core/src/dist/placement.rs` (`Sharded`, `Replicated`, `Partial`), `crates/incin-core/src/dist/plan/` (`HybridPlanner`, `preflight`, `ShardEvidence`), `crates/incin-backends/src/dist/`.

Follows #97. Tensor parallelism additionally depends on #85 and #90. Checkpoint scope line recorded in #105.
